src/rpc/connection.rs

```diff
@@ -675,12 +671,13 @@ impl CircularQueue {
                         "callback data length: {}",
                         (*(callback as *mut OperationCallback)).data_length
                     );
-                    Ok((
-                        (*(callback as *mut OperationCallback)).request_status,
-                        (*(callback as *mut OperationCallback)).flags,
-                        (*(callback as *mut OperationCallback)).meta_data_length,
-                        (*(callback as *mut OperationCallback)).data_length,
-                    ))
+                    let status = (*(callback as *mut OperationCallback)).request_status;
+                    let flags = (*(callback as *mut OperationCallback)).flags;
+                    let meta_data_length = (*(callback as *mut OperationCallback)).meta_data_length;
+                    let data_length = (*(callback as *mut OperationCallback)).data_length;
+
+                    (*(callback as *mut OperationCallback)).state = CallbackState::Done;
```

Comment:
it may be better to set state as Done in response() because clean_up() should happen after setting state.

Reply:
In short, I don't think it should be done.

Reply:
I got it. There is no need to clean up callbacks immediately while we need to ensure the data read completely. 
If response(id1) read the state of id1 as WaitingForResponse and then break, the response(id2) will also clean up the previous callback id1.

Reply:
> I got it. There is no need to clean up callbacks immediately while we need to ensure the data read completely. If clean_up() in response(id1) read the state of id1 as WaitingForResponse and then break, the response(id2) will also clean up the previous callback id1.

Yes, that's the point.